### PR TITLE
#9Task. ivan.lozovskij 

### PR DESCRIPTION
**Pull request by Lozovskij**
https://travis-ci.org/github/Lozovskij/js-assignments/builds/706935352

### CODE REVIEW

File: task/09-functions-n-closures-tasks.js [hunk 1]
```diff
@@ -26,7 +26,10 @@
  *
  */
 function getComposition(f,g) {
-    throw new Error('Not implemented');
+    function res (x) {
+        return f(g(x));
+    }
+    return res;
```
Comment (elizavetaShcherbatykhAis) [inline task/09-functions-n-closures-tasks.js]: You can write it like:
return (x) => f(g(x));

File: task/09-functions-n-closures-tasks.js [hunk 2]
```diff
@@ -47,7 +50,10 @@ function getComposition(f,g) {
  *
  */
 function getPowerFunction(exponent) {
-    throw new Error('Not implemented');
+    function res(x) {
+        return Math.pow(x, exponent)
+    }
+    return res;
 }
```
Comment (elizavetaShcherbatykhAis) [inline task/09-functions-n-closures-tasks.js:51]: Same :)

File: task/09-functions-n-closures-tasks.js [hunk 3]
```diff
@@ -84,10 +101,18 @@ function getPolynom() {
...
+    let cache;
+    let isFirstTime = true;
+    function firstOrCache() {
+        if (isFirstTime) {
+            isFirstTime = false;
+            cache = func(...arguments);//spred operator
+        } 
+        return cache;
+    }
+    return firstOrCache;
```
Comment (elizavetaShcherbatykhAis) [inline task/09-functions-n-closures-tasks.js]: You can do it in 1-2 rows :)

File: task/09-functions-n-closures-tasks.js [hunk 4]
```diff
@@ -132,7 +171,21 @@ function retry(func, attempts) {
...
  */
 function logger(func, logFunc) {
-    throw new Error('Not implemented');
+    /*function res(...x) { //rest param: x[0],...
+        let len = `${JSON.stringify(x)}`.length;
+        logFunc(`${func.name}(${JSON.stringify(x).substring(1, len - 1)}) starts`);
+        let result = func(...x);//spred operator
+        logFunc(`${func.name}(${JSON.stringify(x).substring(1, len - 1)}) ends`);
+        return result;
+    }*/
```
Comment (elizavetaShcherbatykhAis) [inline task/09-functions-n-closures-tasks.js]: Please don't leave any commented out lines

File: task/09-functions-n-closures-tasks.js [hunk 5]
```diff
@@ -150,7 +203,14 @@ function logger(func, logFunc) {
  *   partialUsingArguments(fn, 'a','b','c','d')() => 'abcd'
  */
 function partialUsingArguments(fn) {
-    throw new Error('Not implemented');
+    let args = [...arguments].slice(1);
+    /*function res(...x) {
+        return fn(...args.concat(x));
+    }*/
```
Comment (elizavetaShcherbatykhAis) [inline task/09-functions-n-closures-tasks.js]: Same

**Comment by elizavetaShcherbatykhAis** [comment]
Please check the comments.
I'll mark the PR as draft please click "ready for review" when it will be corrected!
Thank you!

**Comment by Lozovskij** [comment]
https://travis-ci.org/github/Lozovskij/js-assignments/builds/710151215

**Comment by elizavetaShcherbatykhAis** [comment]
Well done! I close PR as accepted

**Pull request closed without merging**